internal/backend/remote-state/azure/auth/clients.go

```diff
@@ -0,0 +1,227 @@
+	}
+	return storageClientFactory.NewAccountsClient(), nil
+}
+
+type StorageAddresses struct {
+	CloudConfig      cloud.Configuration
+	ResourceGroup    string
+	StorageAccount   string
+	StorageContainer string
+	SubscriptionID   string
+	TenantID         string
+}
+
+// NewContainerClientWithSharedKeyCredentialAndKey gets a container client authenticated with
+// a shared Storage Account Access Key, using previously obtained authentication credentials to
+// obtain said key from the Storage Account.
+func NewContainerClientWithSharedKeyCredential(ctx context.Context, names StorageAddresses, authCred azcore.TokenCredential) (*container.Client, error) {
+	containerClient, _, err := NewContainerClientWithSharedKeyCredentialAndKey(ctx, names, authCred)
+	return containerClient, err
+}
+
+func checkNamesForAccessKeyCredentials(names StorageAddresses) error {
+	var diags tfdiags.Diagnostics
+	if names.ResourceGroup == "" {
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			"Resource Group is empty",
+			"In order to obtain a Storage Account Access Key, a resource group is necessary",
+		))
+	}
+	if names.StorageAccount == "" {
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			"Storage Account is empty",
+			"In order to obtain a Storage Account Access Key, a storage account name is necessary",
+		))
+	}
+	if names.SubscriptionID == "" {
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			"Subscription ID is empty",
+			"In order to obtain a Storage Account Access Key, a subscription id is necessary",
+		))
+	}
+	return diags.Err()
+}
+
+// NewContainerClientWithSharedKeyCredentialAndKey gets a container client and shared key
+// that it's authenticated with. This function should only be used for testing and internally within this package.
+func NewContainerClientWithSharedKeyCredentialAndKey(ctx context.Context, names StorageAddresses, authCred azcore.TokenCredential) (*container.Client, string, error) {
+	client := httpclient.New(ctx)
+	// Lookup the key with an account client
+	accountsClient, err := NewStorageAccountsClient(client, authCred, names.CloudConfig, names.SubscriptionID)
+	if err != nil {
+		return nil, "", err
+	}
+	keys, err := accountsClient.ListKeys(ctx, names.ResourceGroup, names.StorageAccount, nil)
+	if err != nil {
+		return nil, "", fmt.Errorf("error listing access keys on the storage account: %w", err)
+	}
+	if len(keys.Keys) == 0 || keys.Keys[0] == nil || keys.Keys[0].Value == nil {
+		return nil, "", errors.New("malformed structure returned from the ListKeys function")
+	}
+
+	storageAccessKey := *keys.Keys[0].Value
+
+	return newContainerClientFromStorageAccessKey(client, names, storageAccessKey)
+}
+
+func CloudConfigFromAddresses(ctx context.Context, environment, metadataHost string) (cloud.Configuration, error) {
+	if metadataHost != "" {
+		return CloudConfigFromMetadataHost(ctx, metadataHost)
+	}
+	switch environment {
+	case "", "public", "global", "canary":
```

Comment:
Where did this list come from?  We should document that.

Also, we have "german" as an option listed in our current backend documentation, has that been deprecated?  I don't see it in the azure sdk anymore.

Reply:
Short answer: [hamilton](https://github.com/manicminer/hamilton/blob/v0.44.0/environments/environments.go#L103-L118) (Link goes to Apache 2.0 licensed code).

Both hamilton and the go-azure-helpers from hashicorp only support public, usgov, and china as possible backends. I really, really looked as hard as I could for "german" but I don't think it actually exists, or, at least, it's not in any code path for the existing azurerm version. I consolidated the two government options (as they don't differ in any meaningful way) and the canary one is also basically the same as public.